plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/FaultTolerantJdbcClient.java

```diff
@@ -0,0 +1,161 @@
+package io.trino.plugin.jdbc;
+
+import io.trino.plugin.jdbc.mapping.IdentifierMapping;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.security.ConnectorIdentity;
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
+import static io.trino.plugin.jdbc.JdbcWriteSessionProperties.getWriteBatchSize;
+import static io.trino.plugin.jdbc.JdbcWriteSessionProperties.isNonTransactionalInsert;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
+
+public abstract class FaultTolerantJdbcClient
```

Comment:
It feels to me that this should not be an extension of `BaseJdbcClient`. I would rather expect the logic you implemented here to go into `BaseJdbcClient`.
And then we choose if "fault-tolerant" or "normal" of execution should be used based on
 * if connector actually support FTE (the flag if it supported may be passed when concrete subclass of `BaseJdbcClient` is created)
 * the state of the session

With current design `MySqlClient` will will always use extra column on INSERTS, even if FTE is not enabled.

Reply:
Also then the `includePageSinkIdColumn` can become a standard static session-access method. Like the methods in `JdbcWriteSessionProperties`.

Reply:
@arhimondr and i felt there wasn't any benefit to branching on whether FTE is enabled. It'd work just as well if FTE is disabled. Also, we wanted to opt-in to this new behavior on a connector-by-connector basis, which is why i created an additional class to extend. Thoughts?

Reply:
Opt-in is still possible without new abstract subclass. YOu may just pass "boolean supportsFTE" when you are calling `super` constructor from `BaseJdbcClient`.
As for branching - IDK - it adds some risk for clusters with FTE disabled, we know old code paths work. The new ones - not so much.

Reply:
That's fair - i can do that

Reply:
just pushed code that makes this change